ledger-core/virtual/execute/delegatedtokenrequest_test.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2020 Insolar Network Ltd.
+// All rights reserved.
+// This material is licensed under the Insolar License version 1.0,
+// available at https://github.com/insolar/assured-ledger/blob/master/LICENSE.md.
+
+package execute
+
+import (
+	"context"
+	"testing"
+
+	"github.com/gojuno/minimock/v3"
+	"github.com/stretchr/testify/require"
+
+	"github.com/insolar/assured-ledger/ledger-core/conveyor/smachine"
+	"github.com/insolar/assured-ledger/ledger-core/conveyor/smachine/smsync"
+	"github.com/insolar/assured-ledger/ledger-core/insolar/contract"
+	"github.com/insolar/assured-ledger/ledger-core/insolar/node"
+	"github.com/insolar/assured-ledger/ledger-core/insolar/payload"
+	"github.com/insolar/assured-ledger/ledger-core/instrumentation/inslogger"
+	"github.com/insolar/assured-ledger/ledger-core/network/messagesender"
+	"github.com/insolar/assured-ledger/ledger-core/network/messagesender/adapter"
+	"github.com/insolar/assured-ledger/ledger-core/pulse"
+	"github.com/insolar/assured-ledger/ledger-core/reference"
+	"github.com/insolar/assured-ledger/ledger-core/testutils/gen"
+	"github.com/insolar/assured-ledger/ledger-core/virtual/object"
+	"github.com/insolar/assured-ledger/ledger-core/virtual/testutils/slotdebugger"
+)
+
+func TestVirtual_CDelegatedCallRequest(t *testing.T) {
+	var (
+		mc       = minimock.NewController(t)
+		ctx      = inslogger.TestContext(t)
+		deadBeef = []byte{0xde, 0xad, 0xbe, 0xef}
+	)
+
+	slotMachine := slotdebugger.New(ctx, t, true)
+	slotMachine.PrepareRunner(ctx, mc)
+
+	var (
+		caller      = gen.UniqueReference()
+		callee      = gen.UniqueReference()
+		outgoing    = gen.UniqueIDWithPulse(slotMachine.PulseSlot.CurrentPulseNumber())
+		objectRef   = reference.NewSelf(outgoing)
+		sharedState = &object.SharedState{
+			Info: object.Info{
+				Reference:      objectRef,
+				PendingTable:   object.NewRequestTable(),
+				KnownRequests:  object.NewRequestTable(),
+				ReadyToWork:    smsync.NewConditional(1, "ReadyToWork").SyncLink(),
+				OrderedExecute: smsync.NewConditional(1, "MutableExecution").SyncLink(),
+			},
+		}
+	)
+	smExecute := SMExecute{
+		Payload: &payload.VCallRequest{
+			CallType:     payload.CTConstructor,
+			CallFlags:    payload.BuildCallFlags(contract.CallTolerable, contract.CallDirty),
+			CallOutgoing: outgoing,
+
+			Caller: caller,
+			Callee: callee,
+		},
+		Meta: &payload.Meta{
+			Sender: caller,
+		},
+	}
+
+	var messageSender adapter.MessageSender = adapter.NewMessageSenderMock(mc).
+		PrepareAsyncMock.Set(
```

Comment:
in slotMachine we already have mocked sender, that could be prepared with
```
slotMachine.PrepareMockedMessageSender(mc)
```
and much easier to use 
```
slotMachine.MessageSender.SendTarget.Set(func(_ context.Context, msg payload.Marshaler, target reference.Global, _ ...messagesender.SendOption) error {
				res := msg.(*payload.VDelegatedCallResponse)
				// ensure that both times request is the same
				assert.Equal(t, caller, target)
				assert.Equal(t, expectedResponse, res)
				return nil
			})
```

Reply:
This way does not allow us to use ExecutionContext for GetPublishedGlobalAliasAndBargeIn and call it

Reply:
the main problem that you generate refIn, it should be static for object and request. So you won't even need to check bargeIn inside call, and just access it with slotMachine.GetPublishedGlobalAliasAndBargeIn, and ensure after step

Reply:
I agree with @sexyhamster , that test construction looks very complicated. I'm sure it can be tested more straightforward way.

Reply:
So now, you don't need to get access to internal context.
You can check that bargeIn is registered after certain step of SMExcecute, get it from PulseSlot and call it in test body, so this code will be simplified

Reply:
Done